README.md
# ARIBA

Antimicrobial Resistance Identification By Assembly

For how to use ARIBA, please see the [ARIBA wiki page][ARIBA wiki].

PLEASE NOTE: we currently do not have the resources to provide support for Ariba - see the [Feedback/Issues](#feedbackissues) section.

[![Unmaintained](http://unmaintained.tech/badge.svg)](http://unmaintained.tech/)  
[![Build Status](https://travis-ci.org/sanger-pathogens/ariba.svg?branch=master)](https://travis-ci.org/sanger-pathogens/ariba)   
[![License: GPL v3](https://img.shields.io/badge/License-GPL%20v3-brightgreen.svg)](https://github.com/sanger-pathogens/ariba/blob/master/LICENSE)   
[![status](https://img.shields.io/badge/MGEN-10.1099%2Fmgen.0.000131-brightgreen.svg)](http://mgen.microbiologyresearch.org/content/journal/mgen/10.1099/mgen.0.000131)   
[![install with bioconda](https://img.shields.io/badge/install%20with-bioconda-brightgreen.svg)](http://bioconda.github.io/recipes/ariba/README.html)  
[![Container ready](https://img.shields.io/badge/container-ready-brightgreen.svg)](https://quay.io/repository/biocontainers/ariba)  
[![Docker Build Status](https://img.shields.io/docker/build/sangerpathogens/ariba.svg)](https://hub.docker.com/r/sangerpathogens/ariba)  
[![Docker Pulls](https://img.shields.io/docker/pulls/sangerpathogens/ariba.svg)](https://hub.docker.com/r/sangerpathogens/ariba)  
[![codecov](https://codecov.io/gh/sanger-pathogens/ariba/branch/master/graph/badge.svg)](https://codecov.io/gh/sanger-pathogens/ariba)

## Contents
* [Introduction](#introduction)
* [Quick Start](#quick-start)
* [Installation](#installation)
  * [Required dependencies](#required-dependencies)
  * [Using pip3](#using-pip3)
  * [From Source](#from-source)
  * [Docker](#docker)
  * [Debian (testing)](#debian-testing)
  * [Ubuntu](#ubuntu)
  * [Dependencies and environment variables](#dependencies-and-environment-variables)
* [Temporary files](#temporary-files)
* [Usage](#usage)
* [License](#license)
* [Feedback/Issues](#feedbackissues)
* [Citation](#citation)

## Introduction
ARIBA is a tool that identifies antibiotic resistance genes by running local assemblies.
It can also be used for [MLST calling](https://github.com/sanger-pathogens/ariba/wiki/MLST-calling-with-ARIBA).

The input is a FASTA file of reference sequences (can be a mix of genes and noncoding sequences) and paired sequencing reads. ARIBA reports which of the reference sequences were found, plus detailed information on the quality of the assemblies and any variants between the sequencing reads and the reference sequences.

## Quick Start
Get reference data, for instance from [CARD](https://card.mcmaster.ca/). See [getref](https://github.com/sanger-pathogens/ariba/wiki/Task%3A-getref) for a full list.

    ariba getref ncbi out.ncbi

Prepare reference data for ARIBA:

    ariba prepareref -f out.ncbi.fa -m out.ncbi.tsv out.ncbi.prepareref

Run local assemblies and call variants:

    ariba run out.ncbi.prepareref reads1.fastq reads2.fastq out.run

Summarise data from several runs:

    ariba summary out.summary out.run1/report1.tsv out.run2/report2.tsv out.run3/report3.tsv

Please read the [ARIBA wiki page][ARIBA wiki] for full usage instructions.

## Tutorials
[The Jupyter notebook tutorial](https://github.com/sanger-pathogens/pathogen-informatics-training)

## Installation

If you encounter an issue when installing ARIBA please contact your local system administrator. If you encounter a bug you can log it [here](https://github.com/sanger-pathogens/ariba/issues).

### Required dependencies
  * [Python3][python] version >= 3.6.0
  * [Bowtie2][bowtie2] version >= 2.1.0
  * [CD-HIT][cdhit] version >= 4.6
  * [MUMmer][mummer] version >= 3.23

ARIBA also depends on several Python packages, all of which are available
via pip. Installing ARIBA with pip3 will get these automatically if they
are not already installed:
  * dendropy >= 4.2.0
  * matplotlib>=3.1.0
  * pyfastaq >= 3.12.0
  * pysam >= 0.9.1
  * pymummer >= 0.10.1
  * biopython

### Using pip3
Install ARIBA using pip:

    pip3 install ariba

### From Source
Download the latest release from this github repository or clone it. Run the tests:

    python3 setup.py test

**Note for OS X:** The tests require gawk which will need to be installed separately, e.g. via Homebrew.

If the tests all pass, install:

    python3 setup.py install

Alternatively, install directly from github using:

    pip3 install git+https://github.com/sanger-pathogens/ariba.git #--user

### Docker
ARIBA can be run in a Docker container. First install Docker, then install ARIBA:

    docker pull sangerpathogens/ariba

To use ARIBA use a command like this (substituting in your directories), where your files are assumed to be stored in /home/ubuntu/data:

    docker run --rm -it -v /home/ubuntu/data:/data sangerpathogens/ariba ariba -h

When calling Ariba via Docker (as above) you'll also need to add **/data/** in front of all the passed in file or directory names (e.g. /data/my_output_folder).


### Debian (Ariba version may not be the latest)
ARIBA is available in the latest version of Debian, and over time will progressively filter through to Ubuntu and other distributions which use Debian. To install it as root:

    sudo apt-get install ariba

### Ubuntu
You can use `apt-get` (see above), or to ensure you get the latest version of ARIBA, the following commands can be
used to install ARIBA and its dependencies. This was tested on a new instance of Ubuntu 16.04.

    sudo  apt-get update
    sudo apt-get install -y python3-dev python3-pip python3-tk zlib1g-dev bowtie2 mummer cd-hit
    export ARIBA_CDHIT=cdhit-est
    sudo pip3 install ariba

### Dependencies and environment variables

By default, ARIBA will look for the dependencies in your `$PATH`, using
the names in the table below. This behaviour can be overridden and
point ARIBA to a specific program using environment variables.
The environment variable is checked first and is used if it is set.
Otherwise ARIBA looks in your `$PATH` for the default name. This applies
to the following dependencies.

| Dependency     |  Default executable    | Environment variable name |
|----------------|------------------------|---------------------------|
| Bowtie2        | `bowtie2`              | `$ARIBA_BOWTIE2`          |
| CD-HIT (est)   | `cd-hit-est`           | `$ARIBA_CDHIT`            |


For example, you could specify an exact version of a bowtie2 executable
that you compiled and downloaded in your home directory (assuming BASH):

    export ARIBA_BOWTIE2=$HOME/bowtie2-2.1.0/bowtie2

Note that ARIBA also runs `bowtie2-build`, for which it uses the
`bowtie2` executable with `-build` appended. So in this case
it would try to use

    $HOME/bowtie2-2.1.0/bowtie2-build

## Temporary files

ARIBA can temporarily make a large number of files whilst running, which
are put in a temporary directory made by ARIBA.  The total size of these
files is small, but there can be a many of them. This can be a
problem when running large numbers (100s or 1000s) of jobs simultaneously
on the same file system.
The parent directory of the temporary directory is determined in the
following order of precedence:

1. The value of the option `--tmp_dir` (if that option was used)
2. The environment variable `$ARIBA_TMPDIR` (if it is set)
3. The environment variable `$TMPDIR` (if it is set)
4. If none of the above is found, then use the run's output directory.

Each temporary directory
is unique to one run of ARIBA, and is automatically deleted at the end
of the run (even if ARIBA was killed by the user or crashed).
For example,

    export $ARIBA_TMPDIR=/tmp

will result in the creation of a new directory inside `/tmp`, which
will have a name of the form

    /tmp/ariba.tmp.abcdef

where the suffix `abcdef` is a random string of characters, chosen
such that `/tmp/ariba.tmp.abcdef` does not already exist.

The exception to the above is if the option `--noclean` is used.
This forces the temporary directory to be placed in the output
directory, and temporary files are kept. It is intended for
debugging.

## Usage
    usage: ariba <command> <options>

    optional arguments:
      -h, --help      show this help message and exit

    Available commands:

	aln2meta      Converts multi-aln fasta and SNPs to metadata
	expandflag    Expands flag column of report file
	flag          Translate the meaning of a flag
	getref        Download reference data
	micplot       Make violin/dot plots using MIC data
	prepareref    Prepare reference data for input to "run"
	pubmlstget    Download species from PubMLST and make db
	pubmlstspecies
		      Get list of available species from PubMLST
	refquery      Get cluster or sequence info from prepareref output
	run           Run the local assembly pipeline
	summary       Summarise multiple reports made by "run"
	test          Run small built-in test dataset
	version       Get versions and exit

Please read the [ARIBA wiki page][ARIBA wiki] for full usage instructions.

## License
ARIBA is free software, licensed under [GPLv3](https://github.com/sanger-pathogens/ariba/blob/master/LICENSE).

## Feedback/Issues
We currently do not have the resources to provide support for Ariba. However, the community might be able to help you out if you report any issues about usage of the software to the [issues page](https://github.com/sanger-pathogens/ariba/issues).

## Citation
If you use this software please cite:

ARIBA: rapid antimicrobial resistance genotyping directly from sequencing reads
Hunt M, Mather AE, Sánchez-Busó L, Page AJ, Parkhill J , Keane JA, Harris SR.
Microbial Genomics 2017. doi: [110.1099/mgen.0.000131](http://mgen.microbiologyresearch.org/content/journal/mgen/10.1099/mgen.0.000131)


  [ariba biorxiv]: http://biorxiv.org/content/early/2017/04/07/118000
  [bowtie2]: http://bowtie-bio.sourceforge.net/bowtie2/index.shtml
  [cdhit]: http://weizhongli-lab.org/cd-hit/
  [ARIBA wiki]: https://github.com/sanger-pathogens/ariba/wiki
  [mummer]: http://mummer.sourceforge.net/
  [python]: https://www.python.org/


CHANGELOG.md
# Change Log

## [v2.14.6](https://github.com/sanger-pathogens/ariba/tree/v2.14.6) (2020-09-07)
[Full Changelog](https://github.com/sanger-pathogens/ariba/compare/v2.14.5...v2.14.6)

**Closed issues:**

- PubMLST URL change breaks pubmlstget [\#304](https://github.com/sanger-pathogens/ariba/issues/304)
- get\_ref card is attempting to download RGI [\#301](https://github.com/sanger-pathogens/ariba/issues/301)
- Reference dataset of megares cannot be downloaded  [\#287](https://github.com/sanger-pathogens/ariba/issues/287)

**Merged pull requests:**

- Fix for changes in PubMLST url [\#305](https://github.com/sanger-pathogens/ariba/pull/305) ([rpetit3](https://github.com/rpetit3))
- Solution 1: for fixing CARD download [\#302](https://github.com/sanger-pathogens/ariba/pull/302) ([rpetit3](https://github.com/rpetit3))
- Update CARD citation message [\#289](https://github.com/sanger-pathogens/ariba/pull/289) ([fmaguire](https://github.com/fmaguire))

## [v2.14.5](https://github.com/sanger-pathogens/ariba/tree/v2.14.5) (2020-05-14)
[Full Changelog](https://github.com/sanger-pathogens/ariba/compare/v2.14.4...v2.14.5)

**Closed issues:**

- Megares Downloading Error [\#294](https://github.com/sanger-pathogens/ariba/issues/294)
- Summarising MLST reports? [\#292](https://github.com/sanger-pathogens/ariba/issues/292)
- CARD 3.0.6 introduces white space in seq id causing fasta/metadata mismatch [\#284](https://github.com/sanger-pathogens/ariba/issues/284)
- ARGannot download fails [\#281](https://github.com/sanger-pathogens/ariba/issues/281)

**Merged pull requests:**

- bump version number [\#297](https://github.com/sanger-pathogens/ariba/pull/297) ([puethe](https://github.com/puethe))
- Handle megares 2.0.0 [\#295](https://github.com/sanger-pathogens/ariba/pull/295) ([martinghunt](https://github.com/martinghunt))
- Correct swapped col and row filters [\#288](https://github.com/sanger-pathogens/ariba/pull/288) ([aunderwo](https://github.com/aunderwo))
- Remove support information [\#286](https://github.com/sanger-pathogens/ariba/pull/286) ([kpepper](https://github.com/kpepper))

## [v2.14.4](https://github.com/sanger-pathogens/ariba/tree/v2.14.4) (2019-09-24)
[Full Changelog](https://github.com/sanger-pathogens/ariba/compare/v2.14.3...v2.14.4)

**Merged pull requests:**

- Fixed GitHub issue \#281 - ARG-ANNOT Fails to Download [\#282](https://github.com/sanger-pathogens/ariba/pull/282) ([kpepper](https://github.com/kpepper))

## [v2.14.3](https://github.com/sanger-pathogens/ariba/tree/v2.14.3) (2019-08-30)
[Full Changelog](https://github.com/sanger-pathogens/ariba/compare/v2.14.2...v2.14.3)

**Closed issues:**

- Version 3.0.3 of CARD breaks prepareref [\#278](https://github.com/sanger-pathogens/ariba/issues/278)
- matplotlib listed twice as a dep in setup.py [\#275](https://github.com/sanger-pathogens/ariba/issues/275)
- v2.14.1	attributeError: 'Namespace' object has no attribute 'cdhit\_max\_memory' [\#273](https://github.com/sanger-pathogens/ariba/issues/273)
- Use ariba run for several isolates in batch [\#262](https://github.com/sanger-pathogens/ariba/issues/262)
- Stopping! Signal received: 13 [\#238](https://github.com/sanger-pathogens/ariba/issues/238)

**Merged pull requests:**

- Added default locale to docker file [\#280](https://github.com/sanger-pathogens/ariba/pull/280) ([kpepper](https://github.com/kpepper))
- Fix for issue \#278 Version 3.0.3 of CARD breaks prepareref [\#279](https://github.com/sanger-pathogens/ariba/pull/279) ([kpepper](https://github.com/kpepper))
- Remove extra matplotlib entry in install\_requires [\#277](https://github.com/sanger-pathogens/ariba/pull/277) ([andersgs](https://github.com/andersgs))
- Removed the specific Ariba release number from the docker file [\#276](https://github.com/sanger-pathogens/ariba/pull/276) ([kpepper](https://github.com/kpepper))

## [v2.14.2](https://github.com/sanger-pathogens/ariba/tree/v2.14.2) (2019-06-21)
[Full Changelog](https://github.com/sanger-pathogens/ariba/compare/v2.14.1...v2.14.2)

**Fixed bugs:**

- Issue with ariba and slurm \(signal 18, signal 20\) [\#263](https://github.com/sanger-pathogens/ariba/issues/263)

**Merged pull requests:**

- RT 660940 - add spades to docker file [\#274](https://github.com/sanger-pathogens/ariba/pull/274) ([kpepper](https://github.com/kpepper))
- Added AUTHORS and LICENSE to the manifest [\#272](https://github.com/sanger-pathogens/ariba/pull/272) ([kpepper](https://github.com/kpepper))

## [v2.14.1](https://github.com/sanger-pathogens/ariba/tree/v2.14.1) (2019-06-13)
[Full Changelog](https://github.com/sanger-pathogens/ariba/compare/v2.14.0...v2.14.1)

**Merged pull requests:**

- Backed out MANIFEST.in change. Bumped release number to v2.14.1. [\#271](https://github.com/sanger-pathogens/ariba/pull/271) ([kpepper](https://github.com/kpepper))

## [v2.14.0](https://github.com/sanger-pathogens/ariba/tree/v2.14.0) (2019-06-06)
[Full Changelog](https://github.com/sanger-pathogens/ariba/compare/v2.13.5...v2.14.0)

**Closed issues:**

- Reference dataset of ARG-ANNOT cannot be downloaded [\#265](https://github.com/sanger-pathogens/ariba/issues/265)
- unable to download mlst schemes [\#264](https://github.com/sanger-pathogens/ariba/issues/264)
- Several "At least one cluster failed!Stopping..." errors [\#261](https://github.com/sanger-pathogens/ariba/issues/261)
- Allow increasing cd-hit-est memory allocation [\#255](https://github.com/sanger-pathogens/ariba/issues/255)
- Ariba pubmlstget Error [\#240](https://github.com/sanger-pathogens/ariba/issues/240)
- segmentation fault when I import ariba [\#230](https://github.com/sanger-pathogens/ariba/issues/230)
- How to use spades and set minimum coverage [\#215](https://github.com/sanger-pathogens/ariba/issues/215)


ariba/report.py
import copy
import re
import sys
import pymummer

class Error (Exception): pass

columns = [
    'ariba_ref_name',        # 0  ariba (renamed) name of reference sequence
    'ref_name',              # 1  original name of ref sequence
    'gene',                  # 2  is a gene 0|1
    'var_only',              # 3  is variant only 0|1
    'flag',                  # 4  cluster flag
    'reads',                 # 5  number of reads in this cluster
    'cluster',               # 6  name of cluster
    'ref_len',               # 7  length of reference sequence
    'ref_base_assembled',    # 8  number of reference nucleotides assembled by this contig
    'pc_ident',              # 9  %identity between ref sequence and contig
    'ctg',                   # 10  name of contig matching reference
    'ctg_len',               # 11  length of contig matching reference
    'ctg_cov',               # 12 mean mapped read depth of this contig
    'known_var',             # 13 is this a known SNP from reference metadata? 1|0
    'var_type',              # 14 The type of variant. Currently only SNP supported
    'var_seq_type',          # 15 if known_var=1, n|p for nucleotide or protein
    'known_var_change',      # 16 if known_var=1, the wild/variant change, eg I42L
    'has_known_var',         # 17 if known_var=1, 1|0 for whether or not the assembly has the variant
    'ref_ctg_change',        # 18 amino acid or nucleotide change between reference and contig, eg I42L
    'ref_ctg_effect',        # 19 effect of change between reference and contig, eg SYS, NONSYN (amino acid changes only)
    'ref_start',             # 20 start position of variant in contig
    'ref_end',               # 21 end position of variant in contig
    'ref_nt',                # 22 nucleotide(s) in contig at variant position
    'ctg_start',             # 23 start position of variant in contig
    'ctg_end',               # 24 end position of variant in contig
    'ctg_nt',                # 25 nucleotide(s) in contig at variant position
    'smtls_total_depth',     # 26 total read depth at variant start position in contig, reported by mpileup
    'smtls_nts',             # 27 alt nucleotides on contig, reported by mpileup
    'smtls_nts_depth',       # 28 alt depth on contig, reported by mpileup
    'var_description',       # 29 description of variant from reference metdata
    'free_text',             # 30 other free text about reference sequence, from reference metadata
]


var_columns = [
    'known_var',
    'var_type',
    'var_seq_type',
    'known_var_change',
    'has_known_var',
    'ref_ctg_change',
    'ref_ctg_effect',
    'ref_start',
    'ref_end',
    'ref_nt',
    'ctg_start',
    'ctg_end',
    'ctg_nt',
    'smtls_total_depth',
    'smtls_nts',
    'smtls_nts_depth',
    'var_description',
]

int_columns = [
    'reads',
    'ref_len',
    'ref_base_assembled',
    'ctg_len',
    'ref_start',
    'ref_end',
    'ctg_start',
    'ctg_end',
]


float_columns = [
    'ctg_cov',
    'pc_ident',
]


def header_line():
    return '\t'.join(columns)


def _samtools_depths_at_known_snps_all_wild(sequence_meta, contig_name, cluster, variant_list):
    '''Input is a known variants, as sequence_metadata object. The
       assumption is that both the reference and the assembly have the
       variant type, not wild type. The list variant_list should be a list
       of pymummer.variant.Variant objects, only contaning variants to the
       relevant query contig'''
    ref_nuc_range = sequence_meta.variant.nucleotide_range()

    if ref_nuc_range is None:
        return None

    bases = []
    ctg_nts = []
    ref_nts = []
    smtls_total_depths = []
    smtls_nts = []
    smtls_depths = []
    contig_positions = []

    for ref_position in range(ref_nuc_range[0], ref_nuc_range[1]+1, 1):
        nucmer_match = cluster.assembly_compare.nucmer_hit_containing_reference_position(cluster.assembly_compare.nucmer_hits, cluster.ref_sequence.id, ref_position, qry_name=contig_name)

        if nucmer_match is not None:
            # work out contig position. Needs indels variants to correct the position
            ref_nts.append(cluster.ref_sequence[ref_position])
            contig_position, in_indel = nucmer_match.qry_coords_from_ref_coord(ref_position, variant_list)
            contig_positions.append(contig_position)
            bases, total_depth, base_depths = cluster.samtools_vars.get_depths_at_position(contig_name, contig_position)
            ctg_nts.append(cluster.assembly.sequences[contig_name][contig_position])
            smtls_nts.append(bases)
            smtls_total_depths.append(total_depth)
            smtls_depths.append(base_depths)

    ctg_nts = ';'.join(ctg_nts) if len(ctg_nts) else '.'
    ref_nts = ';'.join(ref_nts) if len(ref_nts) else '.'
    smtls_nts = ';'.join(smtls_nts) if len(smtls_nts) else '.'
    smtls_total_depths = ';'.join([str(x)for x in smtls_total_depths]) if len(smtls_total_depths) else '.'
    smtls_depths = ';'.join([str(x)for x in smtls_depths]) if len(smtls_depths) else '.'
    ctg_start = str(min(contig_positions) + 1) if contig_positions is not None else '.'
    ctg_end = str(max(contig_positions) + 1) if contig_positions is not None else '.'

    return [str(x) for x in [
        ref_nuc_range[0] + 1,
        ref_nuc_range[1] + 1,
        ref_nts,
        ctg_start,
        ctg_end,
        ctg_nts,
        smtls_total_depths,
        smtls_nts,
        smtls_depths
    ]]


def _report_lines_for_one_contig(cluster, contig_name, ref_cov_per_contig, pymummer_variants):
    lines = []
    reported_known_vars = set()
    contig_length = len(cluster.assembly.sequences[contig_name])
    assert contig_length != 0

    if contig_name in ref_cov_per_contig:
        if contig_name == cluster.assembly_compare.scaff_name_matching_ref and cluster.assembly_compare.gene_matching_ref_type == 'GENE_FOUND':
            ref_cov = len(cluster.ref_sequence)
        else:
            ref_cov = ref_cov_per_contig[contig_name]
    else:
        ref_cov = 0

    common_first_columns = [
        cluster.ref_sequence.id,
        cluster.refdata.ariba_to_original_name.get(cluster.ref_sequence.id, cluster.ref_sequence.id),
        cluster.is_gene,
        cluster.is_variant_only,
        str(cluster.status_flag),
        str(cluster.total_reads),
        cluster.name,
        str(len(cluster.ref_sequence)),
        str(ref_cov),
        str(cluster.assembly_compare.percent_identities[contig_name]) if contig_name in cluster.assembly_compare.percent_identities else '0',
        contig_name,
        str(contig_length),  # 9 length of scaffold matching reference
    ]

    # it's possible that there is no read depth on an assembled contig
    if contig_name in cluster.total_contig_depths:
        common_first_columns.append(str(round(cluster.total_contig_depths[contig_name] / contig_length, 1)))
    else:
        common_first_columns.append('0')

    if cluster.ref_sequence.id in cluster.refdata.metadata and  len(cluster.refdata.metadata[cluster.ref_sequence.id]['.']) > 0:
        free_text_column = ';'.join([x.free_text for x in cluster.refdata.metadata[cluster.ref_sequence.id]['.']])
    else:
        free_text_column = ';'.join(['.'])

    remaining_samtools_variants = copy.copy(cluster.variants_from_samtools)
    if cluster.assembled_ok and contig_name in cluster.assembly_variants and len(cluster.assembly_variants[contig_name]) > 0:
        for (position, var_seq_type, ref_ctg_change, var_effect, contributing_vars, matching_vars_set, metainfo_set) in cluster.assembly_variants[contig_name]:
            if len(matching_vars_set) > 0:
                is_known_var = '1'
                known_var_change = 'unknown'
                var_type = 'SNP'
                has_known_var = '1'
                matching_vars_column = ';;;'.join([x.to_string(separator=':') for x in matching_vars_set])
            else:
                is_known_var = '0'
                known_var_change = '.'
                has_known_var = '0'
                var_type = '.'
                matching_vars_column = '.'

            variant_columns = ['.' if x is None else str(x) for x in [is_known_var, var_type, var_seq_type, known_var_change, has_known_var, ref_ctg_change, var_effect]]

            if contributing_vars is None:
                samtools_columns = [['.'] * 9]
            else:
                if var_effect in ['INDELS', 'MULTIPLE']:
                    ref_start_pos = min([x.ref_start for x in contributing_vars])
                    ref_end_pos = max([x.ref_start for x in contributing_vars])
                    ctg_start_pos = min([x.qry_start for x in contributing_vars])
                    ctg_end_pos = max([x.qry_start for x in contributing_vars])
                else:
                    ref_start_pos = 3 * position if cluster.is_gene == '1' else position
                    assert contig_name in cluster.assembly_compare.nucmer_hits
                    ref_start_hit = None
                    for hit in cluster.assembly_compare.nucmer_hits[contig_name]:
                        if hit.ref_name == cluster.ref_sequence.id and hit.ref_coords().distance_to_point(ref_start_pos) == 0:
                            ref_start_hit = copy.copy(hit)
                            break

                    assert ref_start_hit is not None
                    ctg_start_pos, ctg_start_in_indel = ref_start_hit.qry_coords_from_ref_coord(ref_start_pos, pymummer_variants)

                    if known_var_change not in  ['.', 'unknown']:
                        regex = re.match('^([^0-9]+)([0-9]+)([^0-9]+)$', known_var_change)
                        try:
                            ref_var_string, ref_var_position, ctg_var_string = regex.group(1, 2, 3)
                        except:
                            raise Error('Error parsing variant ' + known_var_change)
                    elif ref_ctg_change != '.':
                        if '_' in ref_ctg_change:
                            regex = re.match('^([^0-9]+)([0-9]+)_[^0-9]+[0-9]+([^0-9]+)$', ref_ctg_change)
                            try:
                                ref_var_string, ref_var_position, ctg_var_string = regex.group(1, 2, 3)
                            except:
                                raise Error('Error parsing variant ' + ref_ctg_change)
                        else:
                            regex = re.match('^([^0-9]+)([0-9]+)([^0-9]+)$', ref_ctg_change)
                            try:
                                ref_var_string, ref_var_position, ctg_var_string = regex.group(1, 2, 3)
                            except:
                                raise Error('Error parsing variant ' + ref_ctg_change)
                    else:
                        assert var_effect == 'SYN'

                    if var_effect == 'SYN':
                        ref_end_pos = ref_start_pos + 2
                        ctg_end_pos = ctg_start_pos + 2
                    elif ref_var_string == '.' or var_effect in {'INS', 'DEL', 'FSHIFT', 'TRUNC', 'INDELS', 'UNKNOWN'}:
                        ref_end_pos = ref_start_pos
                        ctg_end_pos = ctg_start_pos
                    elif cluster.is_gene == '1':
                        ref_end_pos = ref_start_pos + 3 * len(ref_var_string) - 1
                        ctg_end_pos = ctg_start_pos + 3 * len(ctg_var_string) - 1
                    else:
                        ref_end_pos = ref_start_pos + len(ref_var_string) - 1
                        ctg_end_pos = ctg_start_pos + len(ctg_var_string) - 1

                smtls_total_depth = []
                smtls_alt_nt = []
                smtls_alt_depth = []

                for qry_pos in range(ctg_start_pos, ctg_end_pos + 1, 1):
                    if contig_name in remaining_samtools_variants:
                        try:
                            remaining_samtools_variants[contig_name].discard(qry_pos)
                        except:
                            pass

                    depths_tuple = cluster.samtools_vars.get_depths_at_position(contig_name, qry_pos)

                    if depths_tuple is not None:
                        smtls_alt_nt.append(depths_tuple[0])
                        smtls_total_depth.append(str(depths_tuple[1]))
                        smtls_alt_depth.append(str(depths_tuple[2]))

                smtls_total_depth = ';'.join(smtls_total_depth) if len(smtls_total_depth) else '.'
                smtls_alt_nt = ';'.join(smtls_alt_nt) if len(smtls_alt_nt) else '.'
                smtls_alt_depth = ';'.join(smtls_alt_depth) if len(smtls_alt_depth) else '.'
                samtools_columns = [
                        str(ref_start_pos + 1), #ref_start
                        str(ref_end_pos + 1), # ref_end
                        cluster.ref_sequence[ref_start_pos:ref_end_pos+1],
                        str(ctg_start_pos + 1),  # ctg_start
                        str(ctg_end_pos + 1),  #ctg_end
                        cluster.assembly.sequences[contig_name][ctg_start_pos:ctg_end_pos + 1], # ctg_nt
                        smtls_total_depth,
                        smtls_alt_nt,
                        smtls_alt_depth,
                ]


            if len(matching_vars_set) > 0:
                for matching_var in matching_vars_set:
                    if contributing_vars is None:
                        samtools_columns = _samtools_depths_at_known_snps_all_wild(matching_var, contig_name, cluster, pymummer_variants)
                        samtools_columns[2] = samtools_columns[2].replace(';', '')
                        samtools_columns[5] = samtools_columns[5].replace(';', '')
                    reported_known_vars.add(str(matching_var.variant))
                    variant_columns[3] = str(matching_var.variant)

                    if matching_var.has_variant(cluster.ref_sequence) == (ref_ctg_change is not None):
                        variant_columns[4] = '0'
                    else:
                        variant_columns[4] = '1'

                    if samtools_columns is None:
                        samtools_columns = [['.'] * 9]
                    else:
                        for ctg_pos in range(int(samtools_columns[3]) - 1, int(samtools_columns[4]), 1):
                            if contig_name in remaining_samtools_variants:
                                remaining_samtools_variants[contig_name].discard(ctg_pos)

                    lines.append('\t'.join(common_first_columns + variant_columns + samtools_columns + [matching_vars_column] + [free_text_column]))
            else:
                lines.append('\t'.join(
                    common_first_columns + variant_columns + \
                    samtools_columns + \
                    [matching_vars_column] + [free_text_column]
                ))

    if contig_name in remaining_samtools_variants:
        for var_position in remaining_samtools_variants[contig_name]:
            depths_tuple = cluster.samtools_vars.get_depths_at_position(contig_name, var_position)

            if depths_tuple is not None:
                ref_coord, in_indel = None, None
                if contig_name in cluster.assembly_compare.nucmer_hits:
                    for hit in cluster.assembly_compare.nucmer_hits[contig_name]:
                        if hit.qry_coords().distance_to_point(var_position) == 0:
                            ref_coord, in_indel = hit.ref_coords_from_qry_coord(var_position, pymummer_variants)
                            break

                if ref_coord is None:
                    ref_coord = '.'
                    ref_nt = '.'
                    var_string = None
                else:
                    ref_nt = cluster.ref_sequence[ref_coord]
                    ctg_nt = cluster.assembly.sequences[contig_name][var_position]
                    alt_strings = [x for x in depths_tuple[0].split(',') if x != ctg_nt]
                    var_string = ctg_nt + str(ref_coord + 1) + ','.join(alt_strings)
                    ref_coord = str(ref_coord + 1)

                if var_string not in reported_known_vars:
                    new_cols = [
                        '0',  # known_var column
                        'HET', # var_type
                        '.', '.', '.', var_string, '.', ref_coord, ref_coord, ref_nt, # var_seq_type ... ref_nt
                        str(var_position + 1), str(var_position + 1), # ctg_start, ctg_end
                        ctg_nt, # ctg_nt
                        str(depths_tuple[1]), # smtls_total_depth
                        depths_tuple[0], # smtls_alt_nt
                        str(depths_tuple[2]), # smtls_alt_depth
                        '.',
                        free_text_column,
                    ]
                    lines.append('\t'.join(common_first_columns + new_cols))

    if len(lines) == 0:
        lines.append('\t'.join(common_first_columns + ['.'] * (len(columns) - len(common_first_columns) - 1) + [free_text_column]))

    return lines


def report_lines(cluster):
    if cluster.status_flag.has('ref_seq_choose_fail'):
        fields = ['.', '.', '.', '.', str(cluster.status_flag), str(cluster.total_reads), cluster.name] + ['.'] * (len(columns) - 7)
        assert len(fields) == len(columns)
        return ['\t'.join(fields)]
    elif cluster.status_flag.has('assembly_fail'):
        fields = ['.', '.', '.', '.', str(cluster.status_flag), str(cluster.total_reads), cluster.name] + ['.'] * (len(columns) - 7)
        assert len(fields) == len(columns)
        return ['\t'.join(fields)]

    ref_cov_per_contig = cluster.assembly_compare.ref_cov_per_contig(cluster.assembly_compare.nucmer_hits)
    lines = []
    pymummer_variants = pymummer.snp_file.get_all_variants(cluster.assembly_compare.nucmer_snps_file)

    for contig_name in sorted(cluster.assembly.sequences):
        contig_pymummer_variants = [x for x in pymummer_variants if x.qry_name == contig_name]
        lines.extend(_report_lines_for_one_contig(cluster, contig_name, ref_cov_per_contig, contig_pymummer_variants))

    lines_ok = True

    for line in lines:
        if len(line.split('\t')) != len(columns):
            cols = line.split('\t')
            print('Error making report - wrong number of columns. Expected', len(columns), 'but got', len(cols), file=sys.stderr)
            for i in range(len(cols)):
                print(i, cols[i], sep='\t', file=sys.stderr)
            lines_ok = False

    if not lines_ok:
        raise Error('Error making report. Cannot continue')

    return lines if len(lines) > 0 else None

